### DISCUSSION TITLE
Doubt regarding blocking client socket handling

### DISCUSSION THREAD

**Discussion by subhajit-cdot** [strongswan/strongswan#1721] [General]
Hi,
We have implemented a eap exchange based plugin which does external authentication similar to eap-radius.
There is a constraint  that a client socket can handle only one concurrent authentication session at a time that means the interface is blocking and messages are exchanged synchronously, for eg. we have to provide the eap_authenticator SUCCESS/FAILURE/NEED_MORE based on reply received from external authenticator. So there is waiting for the working thread. I want to understand why the design is kept like this? We are hitting a limitation while scaling up this authentication process for multiple user devices. Currently we are using 16 worker thread. Is there any upper limit for number of worker threads in strongswan? What are other design options to handle multiple request ( 500 user device authentication per second ) simultaneously ( if not asynchronously ) with one strongswan instance?

Thanks
Subhajit

**Comment by tobiasbrunner** [comment]
> I want to understand why the design is kept like this?

Because it's very difficult to change.

> Currently we are using 16 worker thread. Is there any upper limit for number of worker threads in strongswan?

Not really, as long as your machine can handle it.

> What are other design options to handle multiple request ( 500 user device authentication per second ) simultaneously ( if not asynchronously ) with one strongswan instance?

If threads are blocked on I/O, you probably need as many as you want to be able to handle clients concurrently. It might also be necessary to drop new IKE_SA_INIT requests or tune job priorities (see [the docs](https://docs.strongswan.org/docs/5.9/config/jobPriority.html)).

**Reply by subhajit-cdot** [reply]
Thanks @tobiasbrunner for the clarification and the document link. Is there any configuration option available to set workery threads to a set of dedicated cores. I know that taskset the process or thread is can be another option but want to know if there is any best practice document available for performance tuning in strongswan . Also pl share if any you have any performance figure or analysis of eap-radius plug-in showing the impact of increasing threads, job priority management with number of concurrent radius request handling , I know it depends on the core’s power, but will there be any other bottleneck or it will scale up linearly?) 

Thanks
Subhajit

**Reply by tobiasbrunner** [reply]
I don't have any such information, sorry.
